crates/matrix-sdk/src/event_cache/room/events.rs

```diff
@@ -129,24 +167,182 @@ impl RoomEvents {
+    /// room_events.remove_events(['a'])
+    ///
+    /// // `position` no longer targets 'b', it now targets 'c', because all
+    /// // items have shifted to the left once. Instead, let's do:
+    ///
+    /// let room_events = room_events(['a', 'b', 'c']);
+    /// let position = position_of('b' in room_events);
+    /// room_events.remove_events_and_update_insert_position(['a'], &mut position)
+    ///
+    /// // `position` has been updated to still target 'b'.
+    /// ```
+    ///
+    /// This is used to remove duplicated events, see
+    /// [`Self::filter_duplicated_events`].
+    fn remove_events_and_update_insert_position(
+        &mut self,
+        event_ids: Vec<OwnedEventId>,
+        position: &mut Position,
+    ) {
+        for event_id in event_ids {
+            let Some(event_position) = self.revents().find_map(|(position, event)| {
+                (event.event_id().as_ref() == Some(&event_id)).then_some(position)
+            }) else {
+                error!(?event_id, "Cannot find te event to remove");
+
+                continue;
+            };
+
+            self.chunks
+                .remove_item_at(event_position)
+                .expect("Failed to remove an event we have just found");
+
+            // A `Position` is composed of a `ChunkIdentifier` and an index.
+            // The `ChunkIdentifier` is stable, i.e. it won't change if an
+            // event is removed in another chunk. It means we only need to
+            // update `position` if the removal happened in **the same
+            // chunk**.
+            if event_position.chunk_identifier() == position.chunk_identifier() {
+                // Now we can compare the position indices.
+                match event_position.index().cmp(&position.index()) {
+                    // `event_position`'s index < `position`'s index
+                    Ordering::Less => {
+                        // An event has been removed _before_ the new
+                        // events: `position` needs to be shifted to the
+                        // left by 1.
+                        position.move_index_to_the_left();
+                    }
+
+                    // `event_position`'s index >= `position`'s index
+                    Ordering::Equal | Ordering::Greater => {
+                        // An event has been removed at the _same_ position of
+                        // or _after_ the new events: `position` does _NOT_ need
+                        // tp be modified.
+                    }
+                }
+            }
+        }
+    }
 }
 
 #[cfg(test)]
 mod tests {
     use assert_matches2::assert_let;
-    use matrix_sdk_test::{EventBuilder, ALICE};
-    use ruma::{events::room::message::RoomMessageEventContent, EventId, OwnedEventId};
+    use ruma::{user_id, EventId, OwnedEventId};
 
     use super::*;
+    use crate::test_utils::events::EventFactory;
+
+    macro_rules! assert_events_eq {
```

Comment:
So much nicer with this macro :star_struck:

Reply:
It is!